[SPARK-5562][MLlib] LDA should handle empty document.

aloknsingh opened the pull request:
See the jira https://issues.apache.org/jira/browse/SPARK-5562

Commit:
[SPARK-5562][MLlib] LDA should handle empty document.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -134,8 +146,10 @@ final class EMLDAOptimizer extends LDAOptimizer {
     }
 
+    // vertices are non empty docs and empty docs
+    val docAllTermVertices = docTermVertices union emptyDocTermVertices
```
I'm not 100% sure this is a hard style rule, but best to avoid this syntax and write `a.union(b)`. It generates a compiler warning otherwise IIRC.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -119,6 +119,18 @@ final class EMLDAOptimizer extends LDAOptimizer {
+    val emptyDocTermVertices: RDD[(VertexId, TopicCounts)] = {
+      //empty documents
+      val emptyDocs = docs.filter { case (d: Long, tc: Vector) =>
+        tc.toBreeze.reduce(_+_) != BDV(0.0)
```
Is this clearer to compute as a norm on `Vector` and just compare to 0 instead of going to breeze and reducing? also I think your case can be like `case (_, tc)`

aloknsingh replied:
agree! in between I realize I had bug with != . it should be == . Since we want empty docs and hence the norm has to be zero.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -119,6 +119,18 @@ final class EMLDAOptimizer extends LDAOptimizer {
+      val emptyDocs = docs.filter { case (d: Long, tc: Vector) =>
+        tc.toBreeze.reduce(_+_) != BDV(0.0)
+      }
+      emptyDocs.map { case(docID: Long, termCounts: Vector) =>
```
You can chain this to avoid more declarations, and seems like you can just use `mapValues`? and would `Vectors.zeros` be simpler and easier?

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala:
```diff
@@ -232,12 +238,20 @@ private[clustering] object LDASuite {
 
+  def getNonEmptyDoc(corpus:Array[(Long, Vector)]): Array[(Long, Vector)] = {
+    corpus.filter { case (docId: Long, wc: Vector) =>
+      wc.toBreeze.reduce(_+_) != BDV(0.0)
```
(Same comment as above, you can avoid a load of declarations here)

aloknsingh replied:
thanks Sean. I will change as per the above comments

Commit:
[SPARK-5562][MLlib] Change as per Sean Owen's comments https://github.com/apache/spark/pull/7064/files#diff-9236d23975e6f5a5608ffc81dfd79146

Commit:
[SPARK-5562][MLlib] change the import xyz._ to the import xyz.{c1, c2} ..

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -119,6 +120,14 @@ final class EMLDAOptimizer extends LDAOptimizer {
+    val emptyDocTermVertices: RDD[(VertexId, TopicCounts)] =
+      docs.filter { case (_, termCounts: Vector) =>
+        Vectors.norm(termCounts, p = 1.0) == 0.0 // empty docs has all zero termCounts
+      }.mapValues { case(termCounts: Vector) =>
```
This can just be `mapValues(termCounts => ...)`. Actually you don't need the type for the `termCounts` above either, or for `emptyDocTermVertices`, I think. That said, maybe don't change this unless you need to do another pass for another reason. And I suppose the code below does write some of this out too. So this seems OK to me as-is as well.

jkbradley replied:
I like specifying the type for emptyDocTermVertices since it's pretty complex, but I agree about the other simplifications.  Style call

SparkQA commented:
[Test build #983 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/983/consoleFull) for   PR 7064 at commit [`9f4f9ea`](https://github.com/apache/spark/commit/9f4f9ea39a99627380cc58cd8f7862893b8191f2).

SparkQA commented:
[Test build #983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/983/console) for   PR 7064 at commit [`9f4f9ea`](https://github.com/apache/spark/commit/9f4f9ea39a99627380cc58cd8f7862893b8191f2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
fix the scala code style to have space after :

aloknsingh commented:
Hi Jenkin , Please test it again.

aloknsingh commented:
Jenkins, retest this please

SparkQA commented:
[Test build #990 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/990/consoleFull) for   PR 7064 at commit [`c710cb6`](https://github.com/apache/spark/commit/c710cb68953bd6a659c8b7abdb0ebd7fde1ac57b).

SparkQA commented:
[Test build #990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/990/console) for   PR 7064 at commit [`c710cb6`](https://github.com/apache/spark/commit/c710cb68953bd6a659c8b7abdb0ebd7fde1ac57b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -119,6 +120,14 @@ final class EMLDAOptimizer extends LDAOptimizer {
       }
     }
 
+    // create term vertices for empty docs
```
These are doc vertices, not term vertices

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -119,6 +120,14 @@ final class EMLDAOptimizer extends LDAOptimizer {
     }
 
+    // create term vertices for empty docs
+    val emptyDocTermVertices: RDD[(VertexId, TopicCounts)] =
```
rename: emptyDocVertices

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala:
```diff
@@ -99,9 +99,15 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
+    // So the output will not contain the empty docs and hence we modify the unittest to
+    // compare against nonEmptyTinyCorpus
+    val nonEmptyTinyCorpus = getNonEmptyDoc(tinyCorpus)
+    assert(topicDistributions.length === nonEmptyTinyCorpus.length)
```
I disagree; there should be topicDistributions for each empty document, probably set to the prior for the corpus.  I think this points to a deeper problem.  I'll comment on that elsewhere.

aloknsingh replied:
As you had pointed out later. going with the options 1 and updating the docs at mllib-clustering.md would be better and I agree with you. I will pull the the changes in.

jkbradley commented:
topicDistributions does not include the empty documents because those vertices are dropped during learning.  Look at this line: [https://github.com/apache/spark/blob/1ce6428907b4ddcf52dbf0c86196d82ab7392442/mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala#L186]

I'd propose 1 of these 2 options:

(1) We simplify things by simply making sure all empty documents are dropped before learning (tested in a unit test).  We should also add documentation in the Scala doc to make it very clear.  The downside is that users who want to join the topicDistributions with the original data will need to filter the original data---but I feel like they should anyways.

(2) Alternatively, we could add the empty document vertices in at the end when creating the DistributedLDAModel.  This might be easier than modifying learning itself.

I'd prefer option (1).  What do you think?

aloknsingh committed:
[SPARK-5562][MLlib] modified the JavaLDASuite for test passing

aloknsingh committed:
[SPARK-5562][Mllib] As per github discussion with jkbradley. We would like to simply things.

Here is the discussion text
topicDistributions does not include the empty documents because those vertices are dropped during learning. Look at this line: [https://github.com/apache/spark/blob/1ce6428907b4ddcf52dbf0c86196d82ab7392442/mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala#L186]

I'd propose 1 of these 2 options:

(1) We simplify things by simply making sure all empty documents are dropped before learning (tested in a unit test). We should also add documentation in the Scala doc to make it very clear. The downside is that users who want to join the topicDistributions with the original data will need to filter the original data---but I feel like they should anyways.

(2) Alternatively, we could add the empty document vertices in at the end when creating the DistributedLDAModel. This might be easier than modifying learning itself.

I'd prefer option (1). What do you think?

aloknsingh committed:
[SPARK-5562][MLlib] fix the newline typo

aloknsingh commented:
Jenkins, retest this please

feynmanliang commented on mllib/src/test/java/org/apache/spark/mllib/clustering/JavaLDASuite.java:
```diff
@@ -33,7 +33,8 @@
 import org.apache.spark.mllib.linalg.Matrix;
 import org.apache.spark.mllib.linalg.Vector;
-
+import org.apache.spark.api.java.function.Function;
```
Sort imports alphabetically

aloknsingh replied:
thanks @feynmanliang . I have put the changes as per comments

Commit:
[SPARK-5562][MLlib] re-order import in alphabhetical order

jkbradley commented on mllib/src/test/java/org/apache/spark/mllib/clustering/JavaLDASuite.java:
```diff
@@ -110,7 +111,16 @@ public void distributedLDAModel() {
+    // SPARK-5562. since the topicDistribution returns the distribution of the non empty docs
+    // over topics. Compare it against nonEmptyCorpus instead of corpus
+    JavaPairRDD<Long, Vector> nonEmptyCorpus = corpus.filter(
+            new Function<Tuple2<Long, Vector>, Boolean>() {
```
Please fix indentation: [https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide]

Actually, I don't think we need to test correctness here since it's tested in the Java unit test

aloknsingh replied:
thanks @jkbradley. I think you meant scala unit test. Please note that the org test was testing for the topicDist vs corpus count. also we test other things like log likelyhood etc in both scala and java unittest.

jkbradley replied:
You're right; I meant scala!  Leaving it sounds fine.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -134,8 +134,9 @@ final class EMLDAOptimizer extends LDAOptimizer {
       verticesTMP.reduceByKey(_ + _)
     }
 
+    val docAllTermVertices = docTermVertices
```
no need for this temp val

jkbradley commented:
Those are the only issues I see.

Commit:
change as per the comments by @jkbradley

aloknsingh commented:
Jenkins, retest this please

SparkQA commented:
[Test build #1000 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1000/consoleFull) for   PR 7064 at commit [`259a0a7`](https://github.com/apache/spark/commit/259a0a7a612004a6a61128ee7611314becb0f0a4).

SparkQA commented:
[Test build #1000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1000/console) for   PR 7064 at commit [`259a0a7`](https://github.com/apache/spark/commit/259a0a7a612004a6a61128ee7611314becb0f0a4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
LGTM merging with master
Thanks!

jkbradley commented:
Apache repo issues...will wait for them to be resolved
